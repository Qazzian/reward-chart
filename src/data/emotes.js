import { combineReducers } from 'redux';

const actionTypes = {
    EMOTE_ADD: 'EMOTE_ADD',
    EMOTE_REMOVE: 'EMOTE_REMOVE',
    EMOTE_UPDATE: 'EMOTE_UPDATE',
};

const emoteTypes = {
    HAPPY: 'HAPPY',
    SAD: 'SAD',
};

function emotesById(state={}, action={}) {
    switch (action.type) {
        case actionTypes.EMOTE_ADD:
            const newKey = generateKey(action.chartId, action.date);
            return {
                ...state,
                [newKey]: emote(state[newKey], action),
            };
        case actionTypes.EMOTE_REMOVE:
            const newState = {...state};
            delete newState[action.emoteId];
            return newState;
        case actionTypes.EMOTE_UPDATE:
            return {
                ...state,
                [action.emoteId]: emote(state[action.emoteId], action),
            };
        default:
            return state;
    }
}

function allIds(state=[], action={}) {
    switch (action.type) {
        case actionTypes.EMOTE_ADD:
            return [
                ...state,
                generateKey(action.chartId, action.date),
            ];
        case actionTypes.EMOTE_REMOVE:
            return state.filter((emoteId) => {
                return emoteId !== action.emoteId;
            });
        default:
            return state;
    }
}

export const emotes = combineReducers({
    emotesById,
    allIds,
});

function emote(state={}, action={}) {
    const {type, ...emoteData} = action;

    const newEmote = {
        ...state,
        ...emoteData,
    };

    const emoteId = newEmote.emoteId || generateKey(newEmote.chartId, newEmote.date);
    return {
        ...newEmote,
        emoteId,
    };
}

// Accessors

export function generateKey(chartId, emoteDate) {
    return `${chartId}_${emoteDate}`;
}

export function getEmoteById(state={}, emoteId) {
    if (state.emotesById) {
        return state.emotesById[emoteId];
    }
    return undefined;
}

export function getEmoteListById(state={}, idList=[]) {
    return idList.map((emoteId) => {
        return state.emotesById[emoteId];
    })
}

// Actions

export function addEmote(chartId, date, emoteType) {
    return {
        type: actionTypes.EMOTE_ADD,
        chartId,
        date,
        emoteType,
    }
}

export function updateEmote(emoteId, emoteType) {
    return {
        type: actionTypes.EMOTE_UPDATE,
        emoteId,
        emoteType,
    };
}

export function removeEmote(emoteId) {
    return {
        type: actionTypes.EMOTE_REMOVE,
        emoteId,
    };
}