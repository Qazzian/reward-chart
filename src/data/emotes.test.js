import * as fromEmotes from './emotes';

const emotes = fromEmotes.emotes;

describe('Emotes state', function () {

    it('should be defined', function () {
        expect(emotes).toBeDefined();
    });

    it('should define the default state', function () {
        const defaultState = {
            emotesById: {},
            allIds: [],
        };
        expect(emotes()).toMatchObject(defaultState);
    });

    it('should add an emote', function () {
        const initState = emotes();
        const action = fromEmotes.addEmote('chart1', '2018-12-14', 'HAPPY');
        const newState = emotes(initState, action);
        expect(newState).toMatchSnapshot();
    });

    it('should update an emote', function () {
        const initState = emotes();
        const addAction = fromEmotes.addEmote('chart1', '2018-12-14', 'HAPPY');
        const testState = emotes(initState, addAction);
        const testAction = fromEmotes.updateEmote('chart1_2018-12-14', 'SAD');
        expect(emotes(testState, testAction)).toMatchSnapshot();
    });

    it('should remove an emote', function () {
        const initState = emotes();
        const addAction1 = fromEmotes.addEmote('chart1', '2018-12-14', 'HAPPY');
        const addAction2 = fromEmotes.addEmote('chart1', '2018-12-15', 'SAD');
        const addAction3 = fromEmotes.addEmote('chart2', '2018-12-14', 'SAD');
        const testState = emotes(emotes(emotes(initState, addAction1), addAction2), addAction3);
        const testAction = fromEmotes.removeEmote('chart1_2018-12-15');
        expect(emotes(testState, testAction)).toMatchSnapshot();
    });
});