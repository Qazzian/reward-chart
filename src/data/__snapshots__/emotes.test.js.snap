// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emotes state should add an emote 1`] = `
Object {
  "allIds": Array [
    "chart1_2018-12-14",
  ],
  "emotesById": Object {
    "chart1_2018-12-14": Object {
      "chartId": "chart1",
      "date": "2018-12-14",
      "emoteId": "chart1_2018-12-14",
      "emoteType": "HAPPY",
    },
  },
}
`;

exports[`Emotes state should get an emote by id 1`] = `
Object {
  "chartId": "chart1",
  "date": "2018-12-14",
  "emoteId": "chart1_2018-12-14",
  "emoteType": "HAPPY",
}
`;

exports[`Emotes state should remove an emote 1`] = `
Object {
  "allIds": Array [
    "chart1_2018-12-14",
    "chart2_2018-12-14",
  ],
  "emotesById": Object {
    "chart1_2018-12-14": Object {
      "chartId": "chart1",
      "date": "2018-12-14",
      "emoteId": "chart1_2018-12-14",
      "emoteType": "HAPPY",
    },
    "chart2_2018-12-14": Object {
      "chartId": "chart2",
      "date": "2018-12-14",
      "emoteId": "chart2_2018-12-14",
      "emoteType": "SAD",
    },
  },
}
`;

exports[`Emotes state should update an emote 1`] = `
Object {
  "allIds": Array [
    "chart1_2018-12-14",
  ],
  "emotesById": Object {
    "chart1_2018-12-14": Object {
      "chartId": "chart1",
      "date": "2018-12-14",
      "emoteId": "chart1_2018-12-14",
      "emoteType": "SAD",
    },
  },
}
`;
